### PR TITLE
Support compound reductions for antialiased lines on the CPU

### LINKED ISSUE #1133: Antialiasing fails on uint32 with a mean reduction
To reproduce, note that the following code works with `line_width=0`:

```python
import datashader as ds
import pandas as pd
cvs = ds.Canvas()
cvs.line(pd.DataFrame({'x': [1, 2, 3], 'y': [2, 3, 4], 'z': np.array([True, False, True]).astype('uint32')},
                      columns=['x', 'y', 'z']), x='x', y='y', agg=ds.mean('z'), line_width=0)
```

But fails with a numba error with a non-zero `line_width`:

<details>
---------------------------------------------------------------------------
TypingError                               Traceback (most recent call last)
Input In [20], in <cell line: 5>()
      1 import datashader as ds
      3 cvs = ds.Canvas()
----> 5 cvs.line(pd.DataFrame({'x': [1, 2, 3], 'y': [2, 3, 4], 'z': np.array([True, False, True]).astype('uint32')},
      6                       columns=['x', 'y', 'z']), x='x', y='y', agg=ds.mean('z'), line_width=1)

File ~/Desktop/development/knauf_panel/stocks_example/envs/default/lib/python3.8/site-packages/datashader/core.py:449, in Canvas.line(self, source, x, y, agg, axis, geometry, line_width, antialias)
    446     # Switch agg to floating point.
    447     agg = rd._reduction_to_floating_point(agg)
--> 449 return bypixel(source, self, glyph, agg)

File ~/Desktop/development/knauf_panel/stocks_example/envs/default/lib/python3.8/site-packages/datashader/core.py:1281, in bypixel(source, canvas, glyph, agg)
   1279 with np.warnings.catch_warnings():
   1280     np.warnings.filterwarnings('ignore', r'All-NaN (slice|axis) encountered')
-> 1281     return bypixel.pipeline(source, schema, canvas, glyph, agg)

File ~/Desktop/development/knauf_panel/stocks_example/envs/default/lib/python3.8/site-packages/datashader/utils.py:109, in Dispatcher.__call__(self, head, *rest, **kwargs)
    107 typ = type(head)
    108 if typ in lk:
--> 109     return lk[typ](head, *rest, **kwargs)
    110 for cls in getmro(typ)[1:]:
    111     if cls in lk:

File ~/Desktop/development/knauf_panel/stocks_example/envs/default/lib/python3.8/site-packages/datashader/data_libraries/pandas.py:17, in pandas_pipeline(df, schema, canvas, glyph, summary)
     15 @bypixel.pipeline.register(pd.DataFrame)
     16 def pandas_pipeline(df, schema, canvas, glyph, summary):
---> 17     return glyph_dispatch(glyph, df, schema, canvas, summary)

File ~/Desktop/development/knauf_panel/stocks_example/envs/default/lib/python3.8/site-packages/datashader/utils.py:112, in Dispatcher.__call__(self, head, *rest, **kwargs)
    110 for cls in getmro(typ)[1:]:
    111     if cls in lk:
--> 112         return lk[cls](head, *rest, **kwargs)
    113 raise TypeError("No dispatch for {0} type".format(typ))

File ~/Desktop/development/knauf_panel/stocks_example/envs/default/lib/python3.8/site-packages/datashader/data_libraries/pandas.py:46, in default(glyph, source, schema, canvas, summary, cuda)
     42 y_axis = canvas.y_axis.compute_index(y_st, height)
     44 bases = create((height, width))
---> 46 extend(bases, source, x_st + y_st, x_range + y_range)
     48 return finalize(bases,
     49                 cuda=cuda,
     50                 coords=OrderedDict([(glyph.x_label, x_axis),
     51                                     (glyph.y_label, y_axis)]),
     52                 dims=[glyph.y_label, glyph.x_label])

File ~/Desktop/development/knauf_panel/stocks_example/envs/default/lib/python3.8/site-packages/datashader/glyphs/line.py:98, in LineAxis0._internal_build_extend.<locals>.extend(aggs, df, vt, bounds, plot_start)
     95     do_extend = extend_cpu
     97 # line may be clipped, then mapped to pixels
---> 98 do_extend(
     99     sx, tx, sy, ty, xmin, xmax, ymin, ymax,
    100     xs, ys, plot_start, *aggs_and_cols
    101 )

File ~/Desktop/development/knauf_panel/stocks_example/envs/default/lib/python3.8/site-packages/numba/core/dispatcher.py:468, in _DispatcherBase._compile_for_args(self, *args, **kws)
    464         msg = (f"{str(e).rstrip()} \n\nThis error may have been caused "
    465                f"by the following argument(s):\n{args_str}\n")
    466         e.patch_message(msg)
--> 468     error_rewrite(e, 'typing')
    469 except errors.UnsupportedError as e:
    470     # Something unsupported is present in the user code, add help info
    471     error_rewrite(e, 'unsupported_error')

File ~/Desktop/development/knauf_panel/stocks_example/envs/default/lib/python3.8/site-packages/numba/core/dispatcher.py:409, in _DispatcherBase._compile_for_args.<locals>.error_rewrite(e, issue_type)
    407     raise e
    408 else:
--> 409     raise e.with_traceback(None)

TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Failed in nopython mode pipeline (step: nopython frontend)
Failed in nopython mode pipeline (step: nopython frontend)
Failed in nopython mode pipeline (step: nopython frontend)
No implementation of function Function(<built-in function imul>) found for signature:
 
 >>> imul(array(uint32, 1d, C), Literal[int](1))
 
There are 8 candidate implementations:
      - Of which 4 did not match due to:
      Overload of function 'imul': File: <numerous>: Line N/A.
        With argument(s): '(array(uint32, 1d, C), int64)':
       No match.
      - Of which 2 did not match due to:
      Overload in function 'NumpyRulesInplaceArrayOperator.generic': File: numba/core/typing/npydecl.py: Line 244.
        With argument(s): '(array(uint32, 1d, C), int64)':
       Rejected as the implementation raised a specific error:
         AttributeError: 'NoneType' object has no attribute 'args'
  raised from /Users/jstevens/Desktop/development/knauf_panel/stocks_example/envs/default/lib/python3.8/site-packages/numba/core/typing/npydecl.py:255
      - Of which 2 did not match due to:
      Operator Overload in function 'imul': File: unknown: Line unknown.
        With argument(s): '(array(uint32, 1d, C), int64)':
       No match for registered cases:
        * (int64, int64) -> int64
        * (int64, uint64) -> int64
        * (uint64, int64) -> int64
        * (uint64, uint64) -> uint64
        * (float32, float32) -> float32
        * (float64, float64) -> float64
        * (complex64, complex64) -> complex64
        * (complex128, complex128) -> complex128

During: typing of intrinsic-call at /Users/jstevens/Desktop/development/knauf_panel/stocks_example/envs/default/lib/python3.8/site-packages/datashader/glyphs/line.py (754)

File "envs/default/lib/python3.8/site-packages/datashader/glyphs/line.py", line 754:
def _full_antialias(line_width, antialias_combination, i, x0, x1, y0, y1,
    <source elided>
    if line_width < 1.0:
        scale *= line_width
        ^

During: resolving callee type: type(CPUDispatcher(<function _full_antialias at 0x7f962e6801f0>))
During: typing of call at <draw_segment> (40)

During: resolving callee type: type(CPUDispatcher(<function _full_antialias at 0x7f962e6801f0>))
During: typing of call at <draw_segment> (40)

File "<draw_segment>", line 40:
<source missing, REPL/exec in use?>

During: resolving callee type: type(CPUDispatcher(<function draw_segment at 0x7f95fb20be50>))
During: typing of call at <perform_extend_line> (21)

During: resolving callee type: type(CPUDispatcher(<function draw_segment at 0x7f95fb20be50>))
During: typing of call at <perform_extend_line> (21)

File "<perform_extend_line>", line 21:
<source missing, REPL/exec in use?>

During: resolving callee type: type(CPUDispatcher(<function perform_extend_line at 0x7f96098000d0>))
During: typing of call at <extend_cpu> (8)

During: resolving callee type: type(CPUDispatcher(<function perform_extend_line at 0x7f96098000d0>))
During: typing of call at <extend_cpu> (8)

File "<extend_cpu>", line 8:
<source missing, REPL/exec in use?>
</details>

This issue crops up when using datashader from HoloViews when trying to datashade boolean data: HoloViews casts to `uint32` as datashader would just complain about the data being non-numeric otherwise.

However, I do think there is an issue here in datasahder: either datashader should explicitly complain that it doesn't support `uint32` or otherwise there shouldn't be cases where errors occur only when antialiasing is enabled.

### PR DESCRIPTION
**Pull request by ianthomas23**
This implements compound reductions such as `mean` and `summary` for antialiased lines on the CPU only. The only reductions not yet covered are `std` and `var`, but there is a possible solution for these being investigated. There is no support for antialiased lines on the GPU, and this will not occur soon as it requires significant changes.

Demonstration of antialiased lines and various reductions:
```python
import datashader as ds
import datashader.transfer_functions as tf
import numpy as np
import pandas as pd

df = pd.DataFrame(dict(ystart=[0.2, 0, 1, 0.8], yend=[0.2, 1, 0, 0.8], value=[1, 2, 3, 4]))

cvs = ds.Canvas(plot_width=200, plot_height=150, x_range=(-0.1, 1.1), y_range=(-0.1, 1.1))
kwargs = dict(source=df, x=np.asarray([0, 1]), y=["ystart", "yend"], axis=1, line_width=15)

for i, reduction in enumerate([
    ds.any(), ds.count(), ds.sum("value"), ds.min("value"), ds.max("value"),
    ds.first("value"), ds.last("value"), ds.mean("value")
]):
    agg = cvs.line(agg=reduction, **kwargs)
    im = tf.shade(agg, how="linear")
    ds.utils.export_image(im, f"temp{i}")
```

Output images for `any`, `count` and `sum`
temp0 temp1 temp2

`min` and `max`
temp3 temp4

`first`, `last`, `mean`
temp5 temp6 temp7

Two observations:

- Note the halos around crossing edges of `min`, `first` and `last`.
- `mean` loses the antialiased edges as it essentially performs a block fill of the pixels that would be rendered partially transparent. This is understandable from the maths of `mean = sum / count`. For a single antialiased line of a constant `value` the pixel values for `sum` equal `count * value`, hence `mean = value`. Note that this is also subject to numerical rounding (not seen here) as typically an antialiased `count` is calculated using `float32` whereas a `sum` uses `float64`.

Closes #1133.

### CODE REVIEW

File: datashader/glyphs/line.py [hunk 1]
```diff
@@ -1107,18 +1152,13 @@ def cpu_antialias_2agg_impl(sx, tx, sy, ty, xmin, xmax, ymin, ymax, xs, ys,
             if ncols == 1:
                 return
 
-            if j == 0:
-                accum_aggs = [aggs[k].copy() for k in range(n_aggs)]
-            else:
-                for k in range(n_aggs):
-                    _combine_in_place(accum_aggs[k], aggs[k], antialias_combinations[k])
+            _aa_stage_2_accumulate(aggs_and_accums, antialias_combinations)
```
Comment (ianthomas23) [inline datashader/glyphs/line.py:1155]: Note the increase in readability of this function and others like it by using separate jit-able functions like `_aa_stage_2_accumulate`. This also allows functions like this to have their `@expand_aggs_and_cols` decorators reinstated, in line with the other various low-level line rendering functions.

File: datashader/compiler.py [hunk 2]
```diff
@@ -59,8 +59,26 @@ def compile_components(agg, schema, glyph, *, antialias=False, cuda=False):
     # List of base reductions (actually computed)
     bases = list(unique(concat(r._build_bases(cuda) for r in reds)))
     dshapes = [b.out_dshape(schema, antialias) for b in bases]
+
+    # Information on how to perform second stage aggregation of antialiased lines,
+    # including whether antialiased lines self-intersect or not as need a single
+    # value for this even for a compound reduction. This is by default True, but
```
Comment (jbednar) [inline datashader/compiler.py:65]:
```suggestion
    # including whether antialiased lines self-intersect or not as we need a single
    # value for this even for a compound reduction. This is by default True, but
```

?

**Review by jbednar — APPROVED** [review]
Thank you for taking this on. My brain wants to crawl into a corner and die when I read this code (before and after this PR, so not your fault!). 

> Note the halos around crossing edges of min, first and last.

I can see those, but can't quite get ahold of the mathematics involved to see if there is any way around that.

> mean loses the antialiased edges as it essentially performs a block fill of the pixels that would be rendered partially transparent. This is understandable from the maths of mean = sum / count. For a single antialiased line of a constant value the pixel values for sum equal count * value, hence mean = value.

Again, I'm not sure if there is a way around this?  Seems like value should be e.g. half the actual value, for the smoothed bits of the edge?

**Comment by ianthomas23** [comment]
> Again, I'm not sure if there is a way around this? Seems like value should be e.g. half the actual value, for the smoothed bits of the edge?

There is no workaround for this, it would need a more complicated implementation as we don't store the information required to do this. We would need to separate out the antialias `weighting` (what would conventionally be called the `alpha` but trying not to use that term here) from the normal aggregate numbers. For the `count` of a single line segment the `count` agg would be 1 for every pixel in the line right up to the edges, and the `weighting` would be 1 along the middle of the line, down to just about 0 at the extreme edges. The 2D agg returned to the user would be the product of the two, giving the same number as we currently return (for a single line).

For a compound reduction of a `count` and `sum` we can imaging 3 aggs, `count`, `sum` and `weighting`. For the `count` we'd return `count*weighting` to the user, for the `sum` we'd return `sum*weighting`, and for the `mean` we'd return `(sum/count)*weighing`, giving you exactly what you want.

In general, we couldn't just have a single `weighting` per `canvas.line` call, we'd need a `weighing` per agg. So maybe we'd think of changing each of the current agg's shapes from `(height, width)` to `(height, width, 2)`, so the agg is attached to its `weighting`.

We would need to decide on appropriate mathematics to combine say `(value1, weighting1)` with `(value2, weighting2)` for a particular pixel. Let's assume we want a linear combination, so the combined `value` must be `value1*weighting1 + value2*weighting2`. Or is it? It might be more sensible to say we want the combined `value` and `weighting` so that `value*weighting = value1*weighting1 + value2*weighting2` because the `value` needs to be stored unweighted (the whole purpose of this approach is to keep the weighting separate from the value). This maths is not dissimilar to combined rendering of RGB and A separately, but there we have the concept of rendering on top of existing colors in a non-commutative way. So it is more complicated than that!

In summary, to obtain a `mean` that more in line with what is expected would require:

1. Twice as much memory for the 3D aggs.
2. Slightly more complicated code, but only at the very highest and lowest levels as we'd still be scanning the `source` dataframe just once, only with 3D aggs.
3. Thought about appropriate maths to combine values and weights.

**Comment by jbednar** [comment]
You say there is no workaround, and then proceed to lay out a workaround. :-) Please open a separate issue proposing that work, emphasizing what limitations antialiasing will have until it is done, but then I don't think there's a particular reason to undertake it at this time. Do mention the limitation in the docs, though, and I guess link to the issue. Thanks!

**Comment by ianthomas23** [comment]
On rereading my previous comment it sounds more positive than I intended it to be! It is not a workaround because it doesn't work; I don't believe there is any maths that can combine `values` and `weights` in a way that works for anything other than the simplest situations. But yes, it is worth writing it up as an issue for further consideration.
